plugins/raiseSkill/raiseSkill.pl

```diff
@@ -0,0 +1,172 @@
+	if ($new_status == INACTIVE) {
+		undef $next_skill;
+		undef @skills_to_add;
+	} elsif ($new_status == ACTIVE) {
+		$active_hooks = Plugins::addHooks(
+			['map_loaded',           \&on_possible_raise_chance],
+			['packet/sendMapLoaded', \&on_possible_raise_chance],
+			['job_level_changed', \&on_possible_raise_chance]
+		);
+	} else {
+		$adding_hook = Plugins::addHooks(
+			['AI_pre',            \&on_ai_pre]
+		);
+	}
+	$status = $new_status;
+}
+
+################################################################
+#  getNextSkill() is the function responsible for deciding
+#  which skill we need to raise next, according to
+#  '@skills_to_add', if we still have skills to raise, set it
+#  to '$next_skill' and return 1, if we have no skills to raise
+#  return 0.
+sub getNextSkill {
+	foreach my $skill (@skills_to_add) {
+		my $char_skill_level = $char->getSkillLevel($skill);
+		my $wanted_skill_level = $skill->getLevel;
+		if ($char_skill_level < $wanted_skill_level) {
+			$next_skill = $skill;
+			return 1;
+		}
+	}
+	message "raiseSkill has no more skills to raise; disabling skillsAddAuto\n", 'success';
+	return 0;
+}
+
+################################################################
+#  canRaise() is the function responsible for checking if
+#  we have free skill points.
+sub canRaise {
+	$char->{points_skill};
+}
+
+################################################################
+#  on_possible_raise_chance() is the function called by
+#  our 'speculative' hooks to try to look for
+#  opportunities to raise skills. It changes the plugin status
+#  to 'ADDING' (2).
+sub on_possible_raise_chance {
+	changeStatus(ADDING) if ($status == ACTIVE);
+}
+
+################################################################
+#  on_ai_pre() is called by 'AI_pre' (duh) when status is
+#  'ADDING' (2), it checks if we can raise our next skill, and
+#  if we can't, change plugin status, otherwise calls raiseSkill()
+sub on_ai_pre {
+	return if !$char;
+	return if $net->getState != Network::IN_GAME;
+	return if !timeOut( $timeout );
+	$timeout->{time} = time;
+	return changeStatus(INACTIVE) unless (getNextSkill());
+	return changeStatus(ACTIVE) unless (canRaise());
+	raiseSkill();
+}
+
+################################################################
+#  raiseSkill() sends to the server our skill raise request and
+#  prints it on console.
+sub raiseSkill {
+	message "Auto-adding skill ".$next_skill->getName." to ".($char->getSkillLevel($next_skill)+1)."\n";
+	$messageSender->sendAddSkillPoint($next_skill->getIDN);
+}
+
+################################################################
+#  on_configModify() is called whenever config is changed, it
+#  checks if something important to us was changed, and can 
+#  change plugin status.
+sub on_configModify {
+	my (undef, $args) = @_;
+	return changeStatus(ACTIVE) if ($args->{key} eq 'skillsAddAuto' && $args->{val} && $config{skillsAddAuto_list} && validateSteps($config{skillsAddAuto_list}));
```

Comment:
I think we should go straight to `ADDING` instead of `ACTIVE`. Let the `AI_pre` hook downgrade back to `ACTIVE` if we can't actually add skills.